From e483cfd30778d74202b892eea719d8f42e805f2b Mon Sep 17 00:00:00 2001
From: Nikita Travkin <nikita@trvn.ru>
Date: Fri, 27 Sep 2019 14:33:13 +0500
Subject: [PATCH 100/210] arm64: dts: qcom: longcheer-l8150: Add custom OTG
 adapter detection

 This patch allows use of custom-made OTG adapters that
 have USB_ID connected to 5V instead of GND.

 USB_ID on l8150 connected to GPIO 37 but pulled down so it's
 impossible to detect normal OTG adapters but custom one
 required anyway since l8150 can't provide power to another
 USB device.

Signed-off-by: Nikita Travkin <nikita@trvn.ru>
---
 .../boot/dts/qcom/msm8916-longcheer-l8150.dts | 31 +++++++++++++++++--
 1 file changed, 29 insertions(+), 2 deletions(-)

diff --git a/arch/arm64/boot/dts/qcom/msm8916-longcheer-l8150.dts b/arch/arm64/boot/dts/qcom/msm8916-longcheer-l8150.dts
index 2df4aecadc5..a25fccec55a 100644
--- a/arch/arm64/boot/dts/qcom/msm8916-longcheer-l8150.dts
+++ b/arch/arm64/boot/dts/qcom/msm8916-longcheer-l8150.dts
@@ -65,6 +65,26 @@ backlight: backlight {
 		default-brightness-level = <255>;
 	};
 
+	usb_otg_hack: usb-otg-hack {
+		compatible = "linux,extcon-usb-gpio";
+
+		/*
+		 * NOTE: This allows detection of custom-made OTG adapters
+		 *
+		 * Since L8150 doesn't support OTG it's impossible to use normal OTG
+		 * adapters. However since USB_ID is in fact connected to GPIO it's
+		 * possible to detect custom-made OTG adapters that connect USB_ID
+		 * to 5v instead of GND and provie external power.
+		 *
+		 * Downside of this approach is that USB_ID pin is used to change boot
+		 * order of SoC and if device will be powered on with adapter attached
+		 * and powered, it will boot into USB BOOT mode.
+		 */
+		id-gpio = <&msmgpio 37 GPIO_ACTIVE_LOW>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&usb_id_default>;
+	};
+
 	gpio-keys {
 		compatible = "gpio-keys";
 
@@ -332,8 +352,7 @@ &sound {
 
 &usb {
 	status = "okay";
-	dr_mode = "peripheral";
-	extcon = <&pm8916_usbin>;
+	extcon = <&pm8916_usbin>, <&usb_otg_hack>;
 };
 
 &usb_hs_phy {
@@ -523,6 +542,14 @@ tp_int_default: tp-int-default {
 		drive-strength = <2>;
 		bias-disable;
 	};
+
+	usb_id_default: usb-id-default {
+		pins = "gpio37";
+		function = "gpio";
+
+		drive-strength = <2>;
+		bias-pull-down;
+	};
 };
 
 &pm8916_mpps {
-- 
2.34.1

